### PR TITLE
feature: posted event queue that delays until the next event loop before executing.

### PR DESCRIPTION
**Pull request by dndx**

### CODE REVIEW

File: patches/nginx-1.11.2-delayed_posted_events.patch [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
++            for (; q != ngx_queue_sentinel(&ngx_posted_delayed_events);
++                 q = ngx_queue_next(q)) {
++                ev = ngx_queue_data(q, ngx_event_t, queue);
++                ev->delayed = 0;
++
++                ngx_log_debug1(NGX_LOG_DEBUG_EVENT, cycle->log, 0,
++                               "skipping delayed posted event %p,"
++                               " til next loop", ev);
++            }
++            break;
```
Comment (agentzh) [inline patches/nginx-1.11.2-delayed_posted_events.patch]: Style: need a blank line before this line.

File: patches/nginx-1.11.2-delayed_posted_events.patch [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     ngx_event_process_posted(cycle, &ngx_posted_events);
++
++    while (!ngx_queue_empty(&ngx_posted_delayed_events)) {
++        q = ngx_queue_head(&ngx_posted_delayed_events);
++
++        ev = ngx_queue_data(q, ngx_event_t, queue);
++        if (ev->delayed) {
++            /* start of newly inserted nodes */
++            for (; q != ngx_queue_sentinel(&ngx_posted_delayed_events);
++                 q = ngx_queue_next(q)) {
```
Comment (agentzh) [inline patches/nginx-1.11.2-delayed_posted_events.patch]: Style: `{` should be on a separate line when the `for` condition is broken into multiple lines.

File: patches/nginx-1.11.2-delayed_posted_events.patch [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
++                ev = ngx_queue_data(q, ngx_event_t, queue);
++                ev->delayed = 0;
++
++                ngx_log_debug1(NGX_LOG_DEBUG_EVENT, cycle->log, 0,
++                               "skipping delayed posted event %p,"
++                               " til next loop", ev);
```
Comment (agentzh) [inline patches/nginx-1.11.2-delayed_posted_events.patch]: "til" => "till"
"next loop" => "next iteration"

File: patches/nginx-1.11.2-delayed_posted_events.patch [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ 
+     ngx_event_process_posted(cycle, &ngx_posted_events);
++
++    while (!ngx_queue_empty(&ngx_posted_delayed_events)) {
++        q = ngx_queue_head(&ngx_posted_delayed_events);
++
++        ev = ngx_queue_data(q, ngx_event_t, queue);
++        if (ev->delayed) {
++            /* start of newly inserted nodes */
++            for (; q != ngx_queue_sentinel(&ngx_posted_delayed_events);
```
Comment (agentzh) [inline patches/nginx-1.11.2-delayed_posted_events.patch]: Better put the stuff after `;` in a separate line to make the code align up more nicely?

Comment (yourenyouyu) [inline patches/nginx-1.11.2-delayed_posted_events.patch]: i think you can provide  ngx.yield()

**Comment by yourenyouyu** [comment]
This solution does not seem to work

```		
               location /lua {
			default_type text/html;
			
			content_by_lua_block {
			while true do
				ngx.sleep(0.001);
			end

			ngx.say("hhh")
			}
		}
		
		location /test {
			default_type text/html;
			
			content_by_lua_block {
			ngx.say("hhh")
			}
		}
```

**Comment by dndx** [comment]
All fixed.

**Comment by agentzh** [comment]
@dndx Merged with the following extra patch:

```diff
diff --git a/patches/nginx-1.11.2-delayed_posted_events.patch b/patches/nginx-1.11.2-delayed_posted_events.patch
index ea636c3..6875843 100644
--- a/patches/nginx-1.11.2-delayed_posted_events.patch
+++ b/patches/nginx-1.11.2-delayed_posted_events.patch
@@ -1,21 +1,18 @@
 diff --git a/src/event/ngx_event.c b/src/event/ngx_event.c
-index c8ae5b2..f59a8ee 100644
+index 57af8132..4853945f 100644
 --- a/src/event/ngx_event.c
 +++ b/src/event/ngx_event.c
-@@ -193,8 +193,10 @@ ngx_module_t  ngx_event_core_module = {
- void
- ngx_process_events_and_timers(ngx_cycle_t *cycle)
- {
--    ngx_uint_t  flags;
--    ngx_msec_t  timer, delta;
-+    ngx_uint_t   flags;
-+    ngx_msec_t   timer, delta;
-+    ngx_queue_t *q;
-+    ngx_event_t *ev;
+@@ -196,6 +196,9 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
+     ngx_uint_t  flags;
+     ngx_msec_t  timer, delta;
  
++    ngx_queue_t     *q;
++    ngx_event_t     *ev;
++
      if (ngx_timer_resolution) {
          timer = NGX_TIMER_INFINITE;
-@@ -215,6 +217,13 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
+         flags = 0;
+@@ -215,6 +218,13 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
  #endif
      }
  
@@ -29,7 +26,7 @@ index c8ae5b2..f59a8ee 100644
      if (ngx_use_accept_mutex) {
          if (ngx_accept_disabled > 0) {
              ngx_accept_disabled--;
-@@ -257,6 +266,35 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
+@@ -257,6 +267,35 @@ ngx_process_events_and_timers(ngx_cycle_t *cycle)
      }
  
      ngx_event_process_posted(cycle, &ngx_posted_events);
@@ -65,7 +62,7 @@ index c8ae5b2..f59a8ee 100644
  }
  
  
-@@ -601,6 +639,7 @@ ngx_event_process_init(ngx_cycle_t *cycle)
+@@ -600,6 +639,7 @@ ngx_event_process_init(ngx_cycle_t *cycle)
  
      ngx_queue_init(&ngx_posted_accept_events);
      ngx_queue_init(&ngx_posted_events);
@@ -74,7 +71,7 @@ index c8ae5b2..f59a8ee 100644
      if (ngx_event_timer_init(cycle->log) == NGX_ERROR) {
          return NGX_ERROR;
 diff --git a/src/event/ngx_event_posted.c b/src/event/ngx_event_posted.c
-index d851f3d..b6cea00 100644
+index d851f3d1..b6cea009 100644
 --- a/src/event/ngx_event_posted.c
 +++ b/src/event/ngx_event_posted.c
 @@ -12,6 +12,7 @@
@@ -86,7 +83,7 @@ index d851f3d..b6cea00 100644
  
  void
 diff --git a/src/event/ngx_event_posted.h b/src/event/ngx_event_posted.h
-index 145d30f..6c38855 100644
+index 145d30fe..6c388553 100644
 --- a/src/event/ngx_event_posted.h
 +++ b/src/event/ngx_event_posted.h
 @@ -43,6 +43,9 @@ void ngx_event_process_posted(ngx_cycle_t *cycle, ngx_queue_t *posted);
diff --git a/util/mirror-tarballs b/util/mirror-tarballs
index 81535db..92c122f 100755
--- a/util/mirror-tarballs
+++ b/util/mirror-tarballs
@@ -57,6 +57,13 @@ fi
 
 answer=`$root/util/ver-ge "$main_ver" 1.11.2`
 if [ "$answer" = "Y" ]; then
+    echo "$info_txt applying the delayed-posted-events patch"
+    patch -p1 < $root/patches/nginx-$main_ver-delayed_posted_events.patch || exit 1
+    echo
+fi
+
+answer=`$root/util/ver-ge "$main_ver" 1.11.2`
+if [ "$answer" = "Y" ]; then
     echo "$info_txt applying the privileged-agent-process patch"
     patch -p1 < $root/patches/nginx-$main_ver-privileged_agent_process.patch || exit 1
     echo
```

**Pull request closed without merging**